include/kernel.h

```diff
@@ -3299,31 +3299,33 @@ extern void k_delayed_work_init(struct k_delayed_work *work,
 /**
  * @brief Submit a delayed work item.
  *
- * This routine schedules work item @a work to be processed by workqueue
- * @a work_q after a delay of @a delay milliseconds. The routine initiates
- * an asynchronous countdown for the work item and then returns to the caller.
- * Only when the countdown completes is the work item actually submitted to
- * the workqueue and becomes pending.
+ * This routine schedules work item @p work to be processed by workqueue @p
+ * work_q after @p delay. The routine schedules a timeout then returns to
+ * the caller.  Only when the timeout completes is the work item actually
+ * submitted to the workqueue and becomes pending.
  *
- * Submitting a previously submitted delayed work item that is still
- * counting down cancels the existing submission and restarts the
- * countdown using the new delay.  Note that this behavior is
- * inherently subject to race conditions with the pre-existing
- * timeouts and work queue, so care must be taken to synchronize such
- * resubmissions externally.
+ * If @p delay is @c K_NO_WAIT then the effect is the same as invoking
+ * k_work_submit_to_queue() after cancelling any incomplete timeout.
+ *
+ * If @p delay is not @c K_NO_WAIT then any incomplete timeout is cancelled
+ * and the work item will be submitted after @p delay.
+ *
+ * @note If a previous delay timeout had completed and the work item is
+ * still pending, but @p delay is not @c K_NO_WAIT, the work item may be
+ * processed twice: once from the previous submit, and once from the new
```

Comment:
> the work item may be processed twice

This would be problematic to deal with in the application. The work item cannot carry any arguments so application has no way to tag it at submission. This way application will have to do some internal refcounting or other tricks.

The ability to cancel timeout and unsubmit was very useful if done in a right context. It allowed to simplify the design.